src/iter/partition.rs

```diff
@@ -0,0 +1,217 @@
+use super::internal::*;
+use super::*;
+
+
+/// Partitions the items of a parallel iterator into a pair of arbitrary
+/// `ParallelExtend` containers.
+pub fn partition<I, C, P>(pi: I, predicate: P) -> (C, C)
+    where I: ParallelIterator,
+          C: Default + ParallelExtend<I::Item>,
+          P: Fn(&I::Item) -> bool + Sync
+{
+    let mut matches = C::default();
+    let mut others = C::default();
+    {
+        // We have no idea what the consumers will look like for these
+        // collections' `par_extend`, but we can intercept them in our own
+        // `drive_unindexed`.  Start with the left side for matching items:
+        let iter = PartitionMatches {
+            base: pi,
+            others: &mut others,
+            predicate: predicate,
+        };
+        matches.par_extend(iter);
+    }
+    (matches, others)
+}
+
+
+/// A fake iterator to intercept the `Consumer` for matching items.
+struct PartitionMatches<'c, I, C: 'c, P> {
+    base: I,
+    others: &'c mut C,
+    predicate: P,
+}
+
+impl<'c, I, C, P> ParallelIterator for PartitionMatches<'c, I, C, P>
+    where I: ParallelIterator,
+          C: Default + ParallelExtend<I::Item>,
+          P: Fn(&I::Item) -> bool + Sync
+{
+    type Item = I::Item;
+
+    fn drive_unindexed<CM>(self, matches: CM) -> CM::Result
+        where CM: UnindexedConsumer<Self::Item>
+    {
+        let mut result = None;
+        {
+            // Now it's time to find the consumer for non-matching items
```

Comment:
Clever trick, I will say. I wonder if we can use some internal traits + generics to "package up" this pattern rather than repeating it. It would probably make things less clear though, although it would provide a good place to document what is going on.